usergrid/usergrid.py

```diff
@@ -1,450 +1,610 @@
-        r.raise_for_status()
-
-        json = r.json()
-
-        self.access_token = json['access_token']
-        self.token_response = json
-        if username:
-            #self.me = json['entities'][0]
-            self.me = json['user']
+        """
+        Login again
 
+        :return:
+        """
+        warnings.warn(DeprecationWarning)
+        self.login()
+
+    @property
     def std_headers(self):
-        headers = {}
-        headers['user-agent'] = 'python usergrid client v.{0}'.format(__version__)
-        headers['Accept'] = 'application/json'
-        headers['Authorization'] = "Bearer {0}".format(self.access_token)
-        if self.use_compression:
+        """
+        Standard headers sent with each request
+
+        :return:
+        """
+        headers = {
+            'user-agent': 'python usergrid client v.{0}'.format(__version__),
+            'Accept': 'application/json',
+        }
+
+        if self._access_token is not None:
+            headers['Authorization'] = "Bearer %s" % self._access_token
+
+        if self._use_compression:
             headers['Accept-Encoding'] = "gzip, deflate"
+
         return headers
 
-    def get_full_endpoint(self, path):
+    def _get_full_endpoint(self, path):
+        """
+        Builds the endpoint for making a request
+
+        :param str path:
+        :rtype str:
+        :return:
+        """
         if '/' in path[0]:
             path = path[1:]
-        full = "{0}/{1}".format(self.app_endpoint, path)
-        return full
-
-    def collect_entities(self, endpoint, ql=None, limit=None):
-        entities = []
-        if not limit or limit < 1000:
-            limit = 1000
-        page_entities, cursor = self.get_entities(endpoint, ql=ql, limit=limit)
-        entities.extend(page_entities)
-        count = len(page_entities)
-
-        old_cursor = None
-        while cursor and (len(page_entities) > 0) and old_cursor != cursor:
-            old_cursor = cursor
-            page_entities, cursor = self.get_entities(
-                endpoint, ql=ql, cursor=cursor, limit=limit)
-            entities.extend(page_entities)
-            count = count + len(page_entities)
-        
 
-        return entities
+        return "{0}/{1}".format(self._app_endpoint, path)
 
-    def process_entities(self, endpoint, method, ql=None, limit=None):
-        page_entities, cursor = self.get_entities(endpoint, ql=ql, limit=limit)
-        # process these entities
-        for entity in page_entities:
-            method(entity)
+    def collect_entities(self, endpoint, ql=None, limit=None):  # pylint: disable=invalid-name
+        """
+        A generator to return all entities
+
+        Do no use this for end-user code
 
-        while cursor:
+        :param str endpoint:
+        :param str ql:
+        :param int limit:
+        :rtype dict:
+        :return:
+        """
+        cursor = None
+
+        if not limit or limit > 100:
+            limit = 100
```

Comment:
collect_entities will page through as long as it has a cursor. There's definitely an advantage to doing this at the UG maximum, 1000 - fewer but larger responses is faster than many smaller responses in adhoc testing over the last 2 years.